elastic/kibana
# [Logstash] [PipelineViewer] Preserve all nested pipeline statements during graph conversion

## Changes
The goal here is to ensure that all of the nested Plugin and If statements of a configuration file are preserved in the Pipeline class when we import the data from the Logstash pipeline graph. Presently there are certain conditions where the existing class can miss nested statements.

Specifically, we've noticed this with `if` statements in the output section; they sometimes show only the first child statement. So an `if` that looks like this:
```
if [condition] {
  stdout { }
  if [condition] {
    file { }
  }
}
```
..will only show the `stdout` plugin in our Pipeline's output section. The `if` sibling statement doesn't show up.

## Testing
There are new unit tests that cover this, but if you'd like to ensure the intended functionality is working

1. Run a pipeline with an if like this in the output section, and monitoring turned on:
```
output {
  if 1 == 1 {
    file {
      path => "/Users/jk/logstash-data/tmp2"
      gzip => true
    }
    if 1 == 1 {
      stdout {
        id => 'secondStandardOut'
      }
    }
  }
}
```
2. Import the pipeline class to the [Pipeline Viewer's directive](https://github.com/justinkambic/kibana/blob/bug-fix/pipeline-class-outputs-show-all-nested-vertices/x-pack/plugins/monitoring/public/directives/logstash/pipeline_viewer/index.js#L12):
`import { Pipeline } from 'plugins/monitoring/components/logstash/pipeline_viewer/models/pipeline';`
3. Add this code to the [Pipeline Viewer's directive](https://github.com/justinkambic/kibana/blob/bug-fix/pipeline-class-outputs-show-all-nested-vertices/x-pack/plugins/monitoring/public/directives/logstash/pipeline_viewer/index.js#L29):
`console.log(Pipeline.fromPipelineGraph(pipelineState.config.graph));`
4. Start up Kibana, navigate to the pipeline monitoring view.
5. In your console, the Pipeline object we created in step 3 should be logged; ensure that the else statements are nested properly.
